transactions/WFLY-17742_SupportTransactionsDuringGracefulShutdown

```diff
@@ -0,0 +1,117 @@
+
+=== Narayana
+
+Narayana will internally handle the lifecycle of transactions during suspension. From the point of view of the integrating party, Narayana should provide a blocking API to suspend itself and it should return control only when there are no transactions left to complete. Moreover, Narayana needs to provide a switch to suspend the creation of new transactions, which needs to be used only when no new transactions are needed.
+
+=== WildFly
+
+At the moment, WildFly’s graceful shutdown cannot be employed in cloud environments out of the box, especially when it comes to handling transactions during its suspension. In fact, as proved with WildFly’s Kubernetes Operator, before scaling down a pod hosting WildFly, all transactions must be completed.
+Even though WildFly’s graceful shutdown already gives the possibility to wait indefinitely, the Transactions subsystem does not really take advantage of this feature. Moreover, there is also an open question about whether the graceful shutdown should be processed synchronously or asynchronously. Currently, the list of subsystems registered with WildFly’s Suspend Controller is processed synchronously, which has its pros and cons.
+
+==== Modifications to WildFly’s Graceful Shutdown (wildfly-core)
+
+Following, modifications to WildFly’s graceful shutdown are discussed in more detail.
+
+*Sync or Async? ->* With regards to the open question presented in the overview, WildFly’s SuspendController synchronously calls pre-suspend and suspend on ServerActivity implementations. As a single timer is started to handle the overall graceful shutdown’s timeout, `ServerActivity` implementations (SAI) have to return very quickly from their pre-suspend/suspend routines to give enough time to other SAIs to process their pre-suspension/suspension. This means that timeout diminishes as SAIs process their suspension. This might result in prematurely cancelling the graceful shutdown for the last SAI in the queue as the overall timeout might have already elapsed. On the other hand, with a synchronous handling, WildFly suspends all registered SAIs one by one, which means that the next SAI does not receive the signal to pre-suspend/suspend until the current SAI has returned the call to pre-suspend/suspend. With an asynchronous suspension, this will no longer be the case and all SAIs will initiate their pre-suspension/suspension concurrently. As a consequence, SAIs should coordinate better their suspensions instead of relying on the intrinsic isolation of the synchronous approach.
+*Proposed modifications.* Introducing an asynchronous graceful shutdown would enable all SAIs to receive a pre-suspend/suspend signal concurrently. In this way, the timeout will have the same duration for all SAIs.
```

Comment:
I'd like to point out here that shutting down the transaction subsystem earlier is just a nice to have because if a resource hasn't been able to recover then there is a good chance it will require some administrative help.